embassy-stm32f4/src/exti.rs

```diff
@@ -0,0 +1,530 @@
+//! Async low power Serial.
+//!
+//! The peripheral is autmatically enabled and disabled as required to save power.
+//! Lowest power consumption can only be guaranteed if the send receive futures
+//! are dropped correctly (e.g. not using `mem::forget()`).
+
+use core::future::Future;
+use core::marker::PhantomData;
+use core::mem;
+use core::pin::Pin;
+
+use embassy::gpio::{WaitForFallingEdge, WaitForRisingEdge};
+use embassy::interrupt::OwnedInterrupt;
+use embassy::util::Signal;
+
+use crate::hal::gpio;
+use crate::hal::gpio::{Edge, ExtiPin as HalExtiPin};
+use crate::hal::syscfg::SysCfg;
+use crate::pac::EXTI;
+
+use crate::interrupt;
+
+pub struct ExtiManager {
+    syscfg: SysCfg,
+}
+
+impl<'a> ExtiManager {
+    pub fn new(_exti: EXTI, syscfg: SysCfg) -> Self {
+        Self { syscfg: syscfg }
+    }
+
+    pub fn new_pin<T, I>(&mut self, mut pin: T, interrupt: I) -> ExtiPin<T, I>
+    where
+        T: HalExtiPin + WithInterrupt<Instance = I>,
+        I: OwnedInterrupt,
+    {
+        pin.make_interrupt_source(&mut self.syscfg);
+
+        ExtiPin {
+            pin: pin,
+            signal: Signal::new(),
+            interrupt: interrupt,
+        }
+    }
+}
+
+pub struct ExtiPin<T: HalExtiPin, I: OwnedInterrupt> {
+    pin: T,
+    signal: Signal<()>,
+    interrupt: I,
+}
+
+/*
+    Irq	Handler	Description
+    EXTI0_IRQn	EXTI0_IRQHandler	Handler for pins connected to line 0
+    EXTI1_IRQn	EXTI1_IRQHandler	Handler for pins connected to line 1
+    EXTI2_IRQn	EXTI2_IRQHandler	Handler for pins connected to line 2
+    EXTI3_IRQn	EXTI3_IRQHandler	Handler for pins connected to line 3
+    EXTI4_IRQn	EXTI4_IRQHandler	Handler for pins connected to line 4
+    EXTI9_5_IRQn	EXTI9_5_IRQHandler	Handler for pins connected to line 5 to 9
+    EXTI15_10_IRQn	EXTI15_10_IRQHandler	Handler for pins connected to line 10 to 15
+*/
+
+impl<T: HalExtiPin, I: OwnedInterrupt> ExtiPin<T, I> {
+    unsafe fn on_interrupt(&self) {
+        self.interrupt.disable();
+        self.signal.signal(());
+    }
+}
+
+impl<T: HalExtiPin, I: OwnedInterrupt> Drop for ExtiPin<T, I> {
+    fn drop(&mut self) {
+        self.interrupt.remove_handler();
```

Comment:
what if the handler is removed, but then the interrupt is not disabled. Does the MCU freeze up?